dm/pkg/checker/conn_checker.go

```diff
@@ -0,0 +1,196 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package checker
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/pingcap/tiflow/dm/config"
+	"github.com/pingcap/tiflow/dm/pkg/conn"
+	tcontext "github.com/pingcap/tiflow/dm/pkg/context"
+	"github.com/pingcap/tiflow/dm/pkg/log"
+	"go.uber.org/zap"
+)
+
+type connAmountChecker struct {
+	toCheckDB *conn.BaseDB
+	stCfgs    []*config.SubTaskConfig
+
+	getConfigConn func(stCfgs []*config.SubTaskConfig) int
+	workerName    string
+	unlimitedConn bool
+}
+
+func newConnAmountChecker(toCheckDB *conn.BaseDB, stCfgs []*config.SubTaskConfig, fn func(stCfgs []*config.SubTaskConfig) int, workerName string) connAmountChecker {
+	return connAmountChecker{
+		toCheckDB:     toCheckDB,
+		stCfgs:        stCfgs,
+		getConfigConn: fn,
+		workerName:    workerName,
+	}
+}
+
+func (c *connAmountChecker) check(ctx context.Context, checkerName string) *Result {
+	result := &Result{
+		Name:  checkerName,
+		Desc:  "check if user-specified number of connection exceeds database's maximum",
+		State: StateFailure,
+	}
+	baseConn, err := c.toCheckDB.GetBaseConn(ctx)
+	//nolint:errcheck
+	defer c.toCheckDB.CloseBaseConn(baseConn)
+	if err != nil {
+		markCheckError(result, err)
+		return result
+	}
+	var rows, processRows *sql.Rows
+	rows, err = baseConn.QuerySQL(tcontext.NewContext(ctx, log.L()), "SHOW GLOBAL VARIABLES LIKE 'max_connections'")
+	if err != nil {
+		markCheckError(result, err)
+		return result
+	}
+	defer rows.Close()
+	var (
+		maxConn  int
+		variable string
+	)
+	for rows.Next() {
+		err = rows.Scan(&variable, &maxConn)
+		if err != nil {
+			markCheckError(result, err)
+			return result
+		}
+	}
+	if maxConn == 0 {
+		c.unlimitedConn = true
+	}
+	processRows, err = baseConn.QuerySQL(tcontext.NewContext(ctx, log.L()), "SHOW PROCESSLIST")
```

Comment:
I remember we can re-use the connection for query before closing previous rows 🤔

Reply:
89e2d9d